### PR TITLE
[network metrics] instance network interface metrics

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Overview
========

This closes the ongoing work related to [plumbing instance/guest metrics](https://github.com/oxidecomputer/propolis/issues/742) through to an oximeter producer in propolis.

Includes:

* An update to `NetworkInterfaceRequest` to include an interface_id, which is part of the request on the [sled instance in Omicron](https://github.com/oxidecomputer/omicron/pull/6414).
* Within initialization, with concrete types, we create interface_ids mapping nic UUIDs => `minor` number of the device instance of viona.
  - The device instance number matches the Kstat we'll look for.
  - We track this mapping via a type alias `InterfaceIdentifiers`, which is passed through from Machine initialization to VM objects (and VM objects shared), the latter of which we can use as part of registering the oximeter producer.
* A new stats module for collecting network_interface metrics.
  - In `to_samples` we generate multiple network interface targets based on vnics per instance, appropriately updating the `interface_id`, which is tracked in `interface_ids`.
* Move kstat_types out of virtual machine-specific stats and use its mocking for other stats tests

Dependencies
------------

- [x] https://github.com/oxidecomputer/omicron/pull/6414
- [x] https://code.illumos.org/c/illumos-gate/+/3630
- [ ] **Note**: Will need to update the commit hash in https://github.com/oxidecomputer/falcon once all this is finished.

### CODE REVIEW

**Comment by zeeshanlakhani** [comment]
This is still awaiting the finalization of work in https://code.illumos.org/c/illumos-gate/+/3630, so it can be tested in practice.

File: Cargo.toml
Comment (zeeshanlakhani) [inline Cargo.toml:171]: keeping this around b/c it's useful for patching.

File: bin/propolis-server/src/lib/stats.rs
Comment (pfmooney) [inline bin/propolis-server/src/lib/stats.rs:42]: Why are the three different stat intervals different?

File: bin/propolis-server/src/lib/stats.rs
Comment (pfmooney) [inline bin/propolis-server/src/lib/stats.rs:217]: PER_CPU limit used for network interfaces?

File: bin/propolis-server/src/lib/stats.rs
Comment (pfmooney) [inline bin/propolis-server/src/lib/stats.rs:221]: This might be a little less tangled if it went something along the lines of:
```rust
let sampler = KstatSampler::with_sample_limit(log, kstat_limit).map_err(|e| {
    slog::error!(
        log,
        "failed to create KstatSampler, \
        vCPU and link stats will be unavailable";
        "error" => ?e,
    );
    e
}).ok()?;
let vcpu_details =
// ...
```

File: bin/propolis-server/src/lib/vm/services.rs
Comment (bnaecker) [inline bin/propolis-server/src/lib/vm/services.rs:54]: If we do land this prior to https://github.com/oxidecomputer/propolis/pull/746, could we copy the interface identifiers here and pass them as owned values? Then we don't need to hold a read lock for the entire time we're setting up the oximeter producer, creating the kstats, etc.

File: crates/propolis-api-types/Cargo.toml
Comment (bnaecker) [inline crates/propolis-api-types/Cargo.toml:12]: I don't see this used anywhere, is this a development change?

File: crates/propolis-api-types/src/lib.rs
Comment (bnaecker) [reply crates/propolis-api-types/src/lib.rs:7]: Ah Patrick beat me to it :)

File: crates/propolis-api-types/src/lib.rs
Comment (bnaecker) [inline crates/propolis-api-types/src/lib.rs:162]: Is this a 2024-edition-preparation thing?

File: crates/rgb-frame/src/lib.rs
Comment (bnaecker) [inline crates/rgb-frame/src/lib.rs:157]: What do these changes do?

File: crates/viona-api/src/lib.rs
Comment (bnaecker) [inline crates/viona-api/src/lib.rs:69]: Maybe a note about why we added this?

File: lib/propolis/src/cpuid.rs
Comment (bnaecker) [inline lib/propolis/src/cpuid.rs:119]: Ok, there are enough of these changes that I'm curious. Did you do them manually, or did you run some `cargo doc` command that fixed them?

**Review by bnaecker — COMMENTED** [review]
Looks pretty good overall, thanks for doing this! I've got some related changes in https://github.com/oxidecomputer/propolis/pull/746 that will conflict, as a head's up. I think merging those would make this change a good bit simpler, so that might be worthwhile. Thanks!

File: bin/propolis-cli/src/main.rs
Comment (zeeshanlakhani) [reply bin/propolis-cli/src/main.rs:34]: yeah. Seemed to be, but since I ended up not touching it, I'll revert.

File: bin/propolis-server/src/lib/initializer.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/initializer.rs:41]: It's because some of the imports were split, and some were not, so I was trying to be consistent.

File: bin/propolis-server/src/lib/stats.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats.rs:42]: They are different. The interface one is similar to what we do with sled metrics vs what was set for vcpu beforehand.

File: rustfmt.toml
Comment (zeeshanlakhani) [reply rustfmt.toml:6]: Yeah. I was the newest Rust and ran into editor issues here. I'll separate this.

File: openapi/propolis-server.json
Comment (zeeshanlakhani) [reply openapi/propolis-server.json:1357]: Yep @pfmooney. This should be done post-merge w/ commit no? I had it in the description.

**Comment by zeeshanlakhani** [comment]
> Looks pretty good overall, thanks for doing this! I've got some related changes in #746 that will conflict, as a head's up. I think merging those would make this change a good bit simpler, so that might be worthwhile. Thanks!

Yeah, I looked at that PR and see how you moved the kstat pieces up. I'm ok updating once that's in.

File: lib/propolis/src/cpuid.rs
Comment (zeeshanlakhani) [reply lib/propolis/src/cpuid.rs:119]: They threw doc errors on `cargo doc` run of CI.

File: crates/rgb-frame/src/lib.rs
Comment (zeeshanlakhani) [reply crates/rgb-frame/src/lib.rs:157]: Escape the slashes. `cargo doc` rendered these with warnings.

File: crates/propolis-api-types/src/lib.rs
Comment (zeeshanlakhani) [reply crates/propolis-api-types/src/lib.rs:162]: Yep. I'll do this in another PR.

File: bin/propolis-server/src/lib/stats/virtual_machine.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/virtual_machine.rs:41]: Yep. I was just `pub(crate)`'ing the new stats module and realized this didn't need to be pub only.

File: crates/propolis-api-types/Cargo.toml
Comment (zeeshanlakhani) [reply crates/propolis-api-types/Cargo.toml:12]: Removed.

File: bin/propolis-server/src/lib/stats/virtual_machine.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/virtual_machine.rs:41]: I'll bring this back to what it was.

Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/virtual_machine.rs:41]: I'll make notes after the updated push.

File: bin/propolis-server/src/lib/vm/services.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/vm/services.rs:54]: I'll bring in #746 to my branch and work with that.

File: bin/propolis-cli/src/main.rs
Comment (zeeshanlakhani) [reply bin/propolis-cli/src/main.rs:34]: Reverted this (on next push).

File: lib/propolis/src/cpuid.rs
Comment (zeeshanlakhani) [reply lib/propolis/src/cpuid.rs:119]: I'll push these out into a separate PR.

File: bin/propolis-server/src/lib/stats.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats.rs:217]: Fix incoming (was an oversight).

File: Cargo.lock
Comment (zeeshanlakhani) [inline Cargo.lock:1]: TODO: I will update the lock correctly once everything from omicron/main and the associated dependencies are in place.

File: Cargo.toml
Comment (zeeshanlakhani) [inline Cargo.toml:99]: Note: This matches omicron/main, as I ran into issues when using omicron/main patched without this update.

File: bin/propolis-server/src/lib/stats/network_interface.rs
Comment (zeeshanlakhani) [inline bin/propolis-server/src/lib/stats/network_interface.rs:51]: I will remove this once tested of course.

File: openapi/propolis-server.json
Comment (zeeshanlakhani) [reply openapi/propolis-server.json:1357]: I've noted this in the PR description.

File: bin/propolis-server/src/lib/stats.rs
Comment (zeeshanlakhani) [inline bin/propolis-server/src/lib/stats.rs:30]: This structure didn't need to be just `pub`. I updated this to be consistent with the updates here in updating modules and making it cleaner than exposing it through the crate (as we already use `mod` here).

**Comment by zeeshanlakhani** [comment]
@pfmooney @bnaecker. Great comments. I removed a bunch of the doc changes and NFCs. I'll move that into another PR. 

@bnaecker I slurpped-in your non-virtual-disk changes from https://github.com/oxidecomputer/propolis/pull/746 into this branch. This made it easier to update and use the newer abstraction(s) for initialization (and avoiding an update to the VM objects structure). We can discuss the best path to order those PRs and whatnot (I'm ok rebasing, w/e works).

**Comment by zeeshanlakhani** [comment]
@pfmooney I think that https://code.illumos.org/c/illumos-gate/+/3630 landed, right? Noob Q: Is the next step an update to Helios?

File: openapi/propolis-server.json
Comment (zeeshanlakhani) [reply openapi/propolis-server.json:1357]: Or, is there a spot where we do that here? Double-checking.

Comment (zeeshanlakhani) [reply openapi/propolis-server.json:1357]: Ah, the openapi `.json` for falcon. Got it. And, it's done manually. Got it.

File: bin/propolis-server/src/lib/stats/network_interface.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/stats/network_interface.rs:115]: nit: s/tragets/targets

File: bin/propolis-server/src/lib/vm/mod.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/vm/mod.rs:144]: I had (am having?) a little trouble following exactly what this type was (or is). It looks like it's an opaque identifier for the kstat series for a particular device; is that right? If so I think I would rename a bit, maybe to `KstatInstanceId`, and update the doc comment to denote that this is just a disambiguating value that might exist for any per-component kstat (it doesn't just have to be a PCI device).

**Review by gjcolombo — COMMENTED** [review]
Thanks for putting this together. The network interface ID plumbing generally makes sense to me, but I had a couple of comments about how things are named/documented that I think could help the concepts to hang together a little bit better. I'll look to do another lap once #746 merges and we rebase this (I think I have mostly kept straight which parts of the PR are from which change but will want to take another look once they're separated out again).

A note about the CI tests--although this is technically an API-breaking change, it's the rare sort of change that I think will *not* break the `phd-run-migrate-from-base` test run, because PHD doesn't currently support guest NICs. If that run does fail once all of the other CI tests are fixed up, we should take a look to make sure we understand what's broken.

File: bin/propolis-server/src/lib/stats.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/stats.rs:51]: @pfmooney `OXIMETER_STAT_INTERVAL` refers to the interval on which we ask `oximeter` to collect data from Propolis. The other intervals are used in the `KstatSampler` to determine how frequently to generate samples from the kstats themselves. Each item registered with the sampler defines its own sampling interval, since there's no one-size-fits-all.

I'm actually going to rename this for clarity in #746 to help address this.

**Comment by pfmooney** [comment]
> @pfmooney I think that https://code.illumos.org/c/illumos-gate/+/3630 landed, right? Noob Q: Is the next step an update to Helios?

It is landed, yes.  The `stlouis` branch periodically does merges from upstream illumos-gate, at which point the new bits would be available in freshly built helios.

File: bin/propolis-server/src/lib/stats.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats.rs:1]: Yeah. This was stats previously, but I can do that in mine PR.

File: crates/propolis-api-types/src/instance_spec/components/devices.rs
Comment (zeeshanlakhani) [reply crates/propolis-api-types/src/instance_spec/components/devices.rs:99]: It's at the [requested_nic](https://github.com/oxidecomputer/omicron/pull/6414/files#diff-e58fdced2e69108a490916cca62b391c44fdde81d564de6f86b8d4e1febc5998R694).

Comment (zeeshanlakhani) [reply crates/propolis-api-types/src/instance_spec/components/devices.rs:99]: I'd say the latter.

Comment (zeeshanlakhani) [reply crates/propolis-api-types/src/instance_spec/components/devices.rs:99]: I updated the comment.

**Comment by zeeshanlakhani** [comment]
Ok, I've rebased against #746, but will wait to push when that's in. I've also got your review comments addressed in my next push @gjcolombo.

**Comment by zeeshanlakhani** [comment]
> > @pfmooney I think that https://code.illumos.org/c/illumos-gate/+/3630 landed, right? Noob Q: Is the next step an update to Helios?
> 
> It is landed, yes. The `stlouis` branch periodically does merges from upstream illumos-gate, at which point the new bits would be available in freshly built helios.

Gotcha. Pinged @citrus-it on this to best get going w/ the testing. Thanks!

**Comment by zeeshanlakhani** [comment]
Just drafting this until #746 is in.

File: bin/propolis-server/src/lib/stats/mod.rs
Comment (pfmooney) [inline bin/propolis-server/src/lib/stats/mod.rs:369]: Why `panic` instead of emitting an error?

This is currently causing `stats::network_interface::test::test_kstat_to_samples` to fail on non-illumos

Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/mod.rs:369]: That was leftover from the previous code, but will update.

Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/mod.rs:369]: And, also update the test :).

Comment (pfmooney) [reply bin/propolis-server/src/lib/stats/mod.rs:369]: The test may pass again if some of this shim code is updated to throw errors rather than panic.

Comment (bnaecker) [reply bin/propolis-server/src/lib/stats/mod.rs]: The `panic!()` was intentional. It looks like the test you're referring to is failing because the code converting from the kstat to the oximeter sample is trying to convert a `NamedData::U32(_)` to a `u64`. We can return an error, but the code is wrong in either case. I wanted to loudly, quickly, and irrecoverably when testing conversion routines that expect some relationship between the types of the kstat and the oximeter sample it generates, in this case, that they're both `u64`s under the hood.

**Comment by bnaecker** [comment]
@zeeshanlakhani Could you avoid force-pushing while the code is under review? GitHub makes it pretty difficult to follow changes in the code as things are overwritten. You can always amend the history as you see fit after the review is done and before merging, either in a new commit or at squash-and-merge time.

**Comment by zeeshanlakhani** [comment]
In live testing with omicron-dev `ch-run`, `oximeter-standalone`, and running `propolis-server` and `propolis-cli` from [these docs], we see the updated time series here, containing `instance_network_interface:*`:

```
instance_network_interface:bytes_received
instance_network_interface:bytes_sent
instance_network_interface:errors_received
instance_network_interface:errors_sent
instance_network_interface:packets_dropped
instance_network_interface:packets_received
instance_network_interface:packets_sent
oximeter_collector:collections
oximeter_collector:failed_collections
virtual_machine:pv_panic_guest_handled
virtual_machine:pv_panic_host_handled
virtual_machine:reset
virtual_machine:vcpu_usage
```

These `instance_network_interface:*` series sample kstats that are available after a recent helios update from the [updating illumos-core from the st. Louis branch](https://github.com/oxidecomputer/illumos-gate/tree/stlouis/) (thank you @pfmooney for the work and @citrus-it for pushing his updates through). These are the representative kstats that are parsed:

```
viona:46:viona_stat:rx_bytes                                                                                                   
viona:46:viona_stat:rx_drops                                                                                                   
viona:46:viona_stat:rx_errors                                                                                                  
viona:46:viona_stat:rx_packets                                                                                                 
viona:46:viona_stat:tx_bytes                                                                                                   
viona:46:viona_stat:tx_drops                                                                                                   
viona:46:viona_stat:tx_errors                                                                                                  
viona:46:viona_stat:tx_packets
``` 

We can check what oximeter's collected with oxql:
```
0x> get instance_network_interface:bytes_received | last 1

instance_id: a5aa5bb1-4fa5-4bb3-8b06-9df335ebc28c
interface_id: 13149b56-8209-493e-b21a-a10367b8ce20
project_id: 33d039a6-d4fc-418b-a544-1f737791d559
silo_id: 6e8d2808-afcc-4d4d-9d05-415b04bff4f4
  [2024-08-29 17:19:55.279446436, 2024-08-29 18:50:18.094871292]: [4846]
```

which matches our `rx_bytes` kstat. 
```
⋊> zeeshan@rainbow ⋊> ~/o/o/o/o/schema on zl/instance-network-interface ⨯ kstat viona:46:viona_stat:rx_bytes           18:50:41
module: viona                           instance: 46
name:   viona_stat                      class:    misc
        rx_bytes                        4846
```

On a different run, let's double-check that the interface_ids match up:

```
network_devices: {
    "vnic-0.5.0": VirtioNic(
        VirtioNic {
            backend_name: "vnic-0.5.0-backend",
            interface_id: ca2dab31-bf60-413c-814c-d1fb78f32736,
            pci_path: PciPath {
                bus: 0,
                device: 5,
                function: 0,
            },
        },
    ),
},...
```
```
0x〉get instance_network_interface:bytes_received | last 1

instance_id: ff4bb3f9-9ce8-431d-a55f-00705f47fb55
interface_id: ca2dab31-bf60-413c-814c-d1fb78f32736
project_id: 1ddbd98a-9bee-42aa-9496-5b2c344e6dbf
silo_id: bea7cb74-1982-4416-8d08-75ffb6854f28
  [2024-08-29 18:57:48.589123332, 2024-08-29 19:01:28.598514660]: [3180]
```

And, yep, they do.

**Comment by zeeshanlakhani** [comment]
> @zeeshanlakhani Could you avoid force-pushing while the code is under review? GitHub makes it pretty difficult to follow changes in the code as things are overwritten. You can always amend the history as you see fit after the review is done and before merging, either in a new commit or at squash-and-merge time.

Yep. I usually do. Apologies as I was keeping the temporary dependencies commit as the recent one.

File: bin/propolis-server/src/lib/stats/mod.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/mod.rs]: Yep. I quickly fixed up the code after changing NamedData type and moved to the error anyhow.

File: bin/propolis-server/src/lib/stats/mod.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/stats/mod.rs:191]: Optional: instead of having bare u32s it might be a bit cleaner to have this function take the instance properties/spec as an argument and peel off the values it needs (instead of giving callers the chance to make a transposition mistake).

File: bin/propolis-server/src/lib/vm/mod.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/vm/mod.rs:115]: Since this is semantically a map, should we just go ahead and make a `BTreeMap` out of it?

File: bin/propolis-server/src/lib/vm/ensure.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/vm/ensure.rs:202]: Instead of having to pass `maybe_kstat_sampler` as an arg to every init function that needs it, I'd strongly consider making it a `MachineInitializer` member at this point. (I think this would involve removing `MachineInitializer::create_kstat_sampler` and then just calling its callee free function directly from here; this shouldn't be too bad since we have all the MachineInitializer members in this function anyway.)

**Review by gjcolombo — COMMENTED** [review]
Apologies for taking so long to do another lap here. This is coming together nicely; most of my feedback is minor style/organizational stuff. The main thing I want to make sure I have right is the description of the new `interface_id` field in the spec API.

File: bin/propolis-server/src/lib/stats/mod.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/stats/mod.rs:148]: Ah no. There were just some inconsistencies around importing it or not. I'll choose the one path in this module.

File: bin/propolis-server/src/lib/spec/config_toml.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/spec/config_toml.rs:296]: Yeah, agreed, as I went back and forth on this. I had `nil` earlier and then went with a generated `v4`.

File: bin/propolis-server/src/lib/vm/mod.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/vm/mod.rs:115]: Yeah, I talked to @bnaecker  on this earlier. Since we only really have 8 or some limited number, it made more sense just to use a Vec and iterate vs a map.

File: bin/propolis-server/src/lib/vm/ensure.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/vm/ensure.rs:202]: Yeah. I had a variation of this earlier before moving the sampler itself into initializer/ensure, as per @bnaecker's PR. I'll do this.

File: bin/propolis-server/src/lib/vm/mod.rs
Comment (zeeshanlakhani) [reply bin/propolis-server/src/lib/vm/mod.rs:115]: I'll add a comment.

**Review by gjcolombo — APPROVED** [review]
Thanks for working in the feedback! This looks good from my end.

File: bin/propolis-server/src/lib/stats/mod.rs
Comment (bnaecker) [inline bin/propolis-server/src/lib/stats/mod.rs:278]: At this point, this UUID is `Uuid::nil()`, right? We probably want to print the list of actual UUIDs, in `nics.interface_ids`, I think.

**Review by bnaecker — APPROVED** [review]
Looks good, thanks for the back and forth! One suggestion about logging, but that's it. Thanks!

**Pull request merged**
